File: core/evidence/semantic_refusal_engine.py
from __future__ import annotations

from typing import Any, Dict, List


def refuse_unsupported_conclusions(
    noninferable_regions: List[str],
    suppressed_speculation: List[Dict[str, Any]],
) -> Dict[str, Any]:
    refusals = []
    for region in noninferable_regions:
        refusals.append({"target": region, "message": "cannot_determine"})
    for spec in suppressed_speculation:
        refusals.append({"target": spec.get("reason", "speculation"), "message": "cannot_determine"})
    return {
        "refusals": refusals,
        "terminated_inferences": [r["target"] for r in refusals],
        "termination_reasons": sorted(set(r["message"] for r in refusals)),
        "unsupported_regions": noninferable_regions,
    }

File: src/evidence/semanticRefusalEngine.ts
/**
 * Converted from Python: core/evidence/semantic_refusal_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function refuseUnsupportedConclusions(noninferable_regions: any, suppressed_speculation: any): any {
  var refusals: any[] = [];
  var region: any;
  for (region of py.iter(noninferable_regions)) {
    py.listAppend(refusals, {"target": region, "message": "cannot_determine"});
  }
  var spec: any;
  for (spec of py.iter(suppressed_speculation)) {
    py.listAppend(refusals, {"target": py.get(spec, "reason", "speculation"), "message": "cannot_determine"});
  }
  return {"refusals": refusals, "terminated_inferences": py.iter(refusals).map((r: any) => py.at(r, "target")), "termination_reasons": py.sorted(py.toSet(py.iter(refusals).map((r: any) => py.at(r, "message")))), "unsupported_regions": noninferable_regions};
}
